### PR TITLE
clean up NEG

### PR DESCRIPTION
**Pull request by freehan**

### CODE REVIEW

**Comment by krzyzacy** [comment]
@freehan 

when I try it locally, I got 
```
- Required 'Alpha Access' permission for 'Compute API'
```

is that something we need to writelist the service account for?

**Comment by freehan** [comment]
oh yeah...

**Comment by krzyzacy** [comment]
```
Try to search for 'network-endpoint-groups' with condition None
['gcloud', 'alpha compute', '-q', 'network-endpoint-groups', 'list', '--format=json(name,creationTimestamp.date(tz=UTC),zone,region,isManaged)', '--filter=name !~ ^default', '--project=k8s-ingress-boskos-10']
ERROR: (gcloud) Invalid choice: 'alpha compute'. Did you mean 'compute'?
Usage: gcloud [optional flags] <group | command>
  group may be           alpha | app | auth | beta | components | compute |
                         config | container | dataflow | dataproc | datastore |
                         debug | deployment-manager | dns | domains |
                         endpoints | firebase | iam | iot | kms | logging | ml |
                         ml-engine | organizations | projects | pubsub |
                         services | source | spanner | sql | topic
  command may be         docker | feedback | help | info | init | version

For detailed information on this command and its flags, run:
  gcloud --help
Fail to list resource 'network-endpoint-groups' from project 'k8s-ingress-boskos-10'

```

Looks like `'alpha compute'` won't work, just make another member for api family I guess...

File: boskos/janitor/janitor.py [hunk 1]
```diff
@@ -79,7 +80,10 @@ def collect(project, age, resource, filt, clear_all):
     if resource.name == 'sinks' and not clear_all:
         return col
 
-    cmd = ['gcloud', resource.group, '-q', resource.name]
+    cmd = ['gcloud']
+    if len(resource.api_version) > 0:
```
Comment (krzyzacy) [inline boskos/janitor/janitor.py]: nit: just `if resource.api_version:`

**Comment by krzyzacy** [comment]
also add the api_version to `clear_resources` call

File: boskos/janitor/janitor.py (hunk 1 above)
Comment (freehan) [reply boskos/janitor/janitor.py]: Done

**Comment by krzyzacy** [comment]
```
gcloud alpha compute network-endpoint-groups list --project=k8s-ingress-boskos-12 --format='json(name,zone)'
[
  {
    "name": "k8s1-f58a6b44-e2e-tests-ingress-865wj-hostname-9376-cb6e9557"
  }
]
```
looks like there's no zone info from the list api, which is necessary for deleting the neg resource
/hold

File: boskos/janitor/janitor.py [hunk 2]
```diff
@@ -168,6 +175,11 @@ def clear_resources(project, cols, resource, rate_limit):
             else:
                 base.append('--global')
 
+        # hard code asia-southeast1-a for NEG
+        # TODO: remove this once limitation is dropped
```
Comment (krzyzacy) [inline boskos/janitor/janitor.py]: nit: `TODO(freehan):`

and please open an issue to track this

Comment (freehan) [reply boskos/janitor/janitor.py]: okay

**Comment by krzyzacy** [comment]
also resolve pylint.. I guess we can make the command construction part a subfunction

**Comment by freehan** [comment]
Filed issue for removing hard code
https://github.com/kubernetes/test-infra/issues/8327

File: boskos/janitor/janitor.py [hunk 3]
```diff
@@ -27,36 +27,50 @@
...
+
+    Args:
+        api_version: the api version of the gcloud.
+    Returns:
+        list of base commands of gcloud .
+    """
+
+    base = ['gcloud']
+    if api_version:
+        cmd += [api_version]
```
Comment (krzyzacy) [inline boskos/janitor/janitor.py]: base

Comment (krzyzacy) [reply boskos/janitor/janitor.py]: probably worth take in resource, so we can have group, name and subgroup all handled here

File: boskos/janitor/janitor.py [hunk 4]
```diff
@@ -27,41 +27,58 @@
...
-    Resource('logging', 'sinks', None, None, None, False, False),
+    Resource('', 'logging', 'sinks', None, None, None, False, False),
 ]
 
 def log(message):
     """ print a message if --verbose is set. """
     if ARGS.verbose:
         print message
 
+def base_command(api_version):
```
Comment (krzyzacy) [inline boskos/janitor/janitor.py]: s/api_version/resource

**Comment by krzyzacy** [comment]
ping ^^

there ~9 ingress projects are stuck

File: boskos/janitor/janitor.py [hunk 5]
```diff
@@ -27,41 +27,58 @@
...
+
+    Args:
+        api_version: the api version of the gcloud.
+    Returns:
+        list of base commands of gcloud .
+    """
+
+    base = ['gcloud']
+    if api_version:
+        base += [api_version]
```
Comment (krzyzacy) [inline boskos/janitor/janitor.py]: resource.api_version

File: boskos/janitor/janitor.py [hunk 6]
```diff
@@ -27,41 +27,58 @@
...
 def log(message):
     """ print a message if --verbose is set. """
     if ARGS.verbose:
         print message
 
+def base_command(resource):
+    """ Return the base gcloud command with api_version if exists.
+
+    Args:
+        api_version: the api version of the gcloud.
```
Comment (krzyzacy) [inline boskos/janitor/janitor.py]: update docstring :-)

File: boskos/janitor/janitor.py [hunk 7]
```diff
@@ -247,6 +262,12 @@ def clean_gke_cluster(project, age, filt):
 
     return err
 
+# hard code asia-southeast1-a for NEG
+# TODO(freehan): remove this once limitation is dropped
+def inject_zone_for_neg(resource_name):
+    if resource_name == 'network-endpoint-groups':
+        return ["--zone=asia-southeast1-a"]
```
Comment (krzyzacy) [inline boskos/janitor/janitor.py]: last nit: single quote, as everywhere else

**Comment by krzyzacy** [comment]
/lgtm
/hold
works locally

File: boskos/janitor/janitor.py [hunk 8]
```diff
@@ -247,6 +262,12 @@ def clean_gke_cluster(project, age, filt):
 
     return err
 
+# hard code asia-southeast1-a for NEG
+# TODO(freehan): remove this once limitation is dropped
+def inject_zone_for_neg(resource_name):
+    if resource_name == 'network-endpoint-groups':
+        return [`--zone=asia-southeast1-a`]
```
Comment (krzyzacy) [inline boskos/janitor/janitor.py]: single quote, not back tick :-)))))

**Pull request merged**